src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs

```diff
@@ -263,57 +266,98 @@ public static async Task<PackageSpec> GetProjectSpec(IDependencyGraphProject pro
             DependencyGraphCacheContext context)
         {
             var dgSpec = new DependencyGraphSpec();
-            List<IAssetsLogMessage> allAdditionalMessages = null;
+            ConcurrentBag<IAssetsLogMessage> allAdditionalMessages = null;
 
             var projects = (await solutionManager.GetNuGetProjectsAsync()).OfType<IDependencyGraphProject>().ToList();
-            var knownProjects = projects.Select(e => e.MSBuildProjectPath).ToHashSet(PathUtility.GetStringComparerBasedOnOS());
+            var knownProjects = new ConcurrentDictionary<string, bool>(PathUtility.GetStringComparerBasedOnOS());
+            knownProjects.AddRange(projects.Select(e => e.MSBuildProjectPath)
+                .Select(proj => new KeyValuePair<string, bool>(proj, false)));
 
-            for (var i = 0; i < projects.Count; i++)
+            var options = new ExecutionDataflowBlockOptions()
             {
-                var (packageSpecs, projectAdditionalMessages) = await projects[i].GetPackageSpecsAndAdditionalMessagesAsync(context);
+                MaxDegreeOfParallelism = Environment.ProcessorCount
+            };
+            var actionBlock = new ActionBlock<ProjectRestoreSpec>(GetProjectRestoreSpecAndAdditionalMessages, options);
 
-                if (projectAdditionalMessages != null && projectAdditionalMessages.Count > 0)
+            foreach (IDependencyGraphProject project in projects)
+            {
+                var (packageSpecs, projectAdditionalMessages) = await project.GetPackageSpecsAndAdditionalMessagesAsync(context);
+                await actionBlock.SendAsync(new ProjectRestoreSpec()
                 {
-                    if (allAdditionalMessages == null)
-                    {
-                        allAdditionalMessages = new List<IAssetsLogMessage>();
-                    }
+                    DgSpec =  dgSpec,
+                    PackageSpecs = packageSpecs,
+                    ProjectAdditionalMessages = projectAdditionalMessages,
+                    KnownProjects = knownProjects,
+                    AllAdditionalMessages = allAdditionalMessages
+                });
+            }
+
+            actionBlock.Complete();
+            await actionBlock.Completion;
+
+            // Return dg file
+            return (dgSpec, allAdditionalMessages?.ToList());
+        }
+
+        private static void GetProjectRestoreSpecAndAdditionalMessages(
+            ProjectRestoreSpec restoreSpecData)
+        {
+            var dgSpec = restoreSpecData.DgSpec;
+
+            if (restoreSpecData.ProjectAdditionalMessages?.Any() ?? false)
+            {
+                if (restoreSpecData.AllAdditionalMessages == null)
+                {
+                    restoreSpecData.AllAdditionalMessages = new ConcurrentBag<IAssetsLogMessage>();
```

Comment:
This initialization is not thread safe. Can it be initialized before being passed to the restoreSpecData and this code segment to be removed?

Reply:
Ok. Done.